[WIP] Add PDF annotation

aswinmprabhu opened the pull request.

oliversauter commented:
1. Fingerprinting does not work properly yet
    1. I you add just a comment in the sidebar here: [http://comm.stanford.edu/mm/2016/07/Marisa-Messsina-MA-Thesis.pdf](http://comm.stanford.edu/mm/2016/07/Marisa-Messsina-MA-Thesis.pdf), then go to "annotate pdf" it does not show the comment in the chrome-extension://abc addressed PDF. 
2. Fetching and storing of annotations does not work yet:
   - [ ] When trying to save an annotation, following errors: <img width="839" alt="screen shot 2019-02-01 at 12 38 50" src="https://user-images.githubusercontent.com/7870039/52164312-ab085500-26ef-11e9-8a7d-7623cde5687a.png">
   - [ ] On page load of the [url adressed PDF](http://comm.stanford.edu/mm/2016/07/Marisa-Messsina-MA-Thesis.pdf) these are the errors: <img width="840" alt="screen shot 2019-02-01 at 12 46 49" src="https://user-images.githubusercontent.com/7870039/52164328-ef93f080-26ef-11e9-90f1-c488ac0bbb2f.png">
   - [ ] Somtimes when clicking on the small arrow of the ribbon it does not open the sidebar, instead showing these errors with each click on the arrow button:  <img width="842" alt="screen shot 2019-02-01 at 12 50 37" src="https://user-images.githubusercontent.com/7870039/52164334-00dcfd00-26f0-11e9-9a0d-8191f6c28ed5.png">
   - [ ] Sometimes the comments made on the url version do not show, and infinite loading bar. Also it should check with the fingerprint if it is a pdf we have already. 
   - [ ] We need the context menu for the annotations
    it is not very intuitive on how to start annotating a PDF so we need 3 mechanisms:
      - [ ] Opening the ribbon on a PDF, immediately puts it in annotation mode (otherwise the highlights cannot be shown)
      - [ ] enabling highlights (A feature currently in the making) with the ribbon highlighter button should also trigger annotation mode. 
      - [ ] Context menu (right click on highlighted text) to start annotating and kicking the PDF into annotation mode
      - [ ] A small button in the sidebar enabling annotating the pdf (similar to the popup)

aswinmprabhu commented:
@oliversauter The comments and annotations are not supposed to work. I have not started working on them yet. So the errors are a bit meaningless at this point. I have been trying to fix the sidebar UI and getting the content script to work in the PDF.js viewer. Now that its working, I will start working on comments and annotation. I will start work on your pdf-annotation workflow suggestions as well.

oliversauter commented:
> Now that its working, I will start working on comments and annotation.

Ok thanks for clearing this up. Great job so far. Also appreciate your clear and continuous communication!

poltak commented on build/config.js:
```diff
@@ -60,6 +60,9 @@ export default ({ context = __dirname, mode = 'development', ...opts }) => {
             hints: false,
         },
+        node: {
+            fs: 'empty',
```
Can you give me some more info as to why this is needed? Or, what expects node's `fs` module to be available?

aswinmprabhu replied:
![image](https://user-images.githubusercontent.com/31558262/52568378-9b49e880-2e34-11e9-88d2-b12c3b19c99f.png)
Without it, I am getting this webpack build error. A quick google search gave me this solution.

poltak commented on build/static-files.js:
```diff
@@ -32,7 +32,7 @@ export const copyPatterns = [
     { from: 'img', to: 'img' },
     {
-        from: 'node_modules/webextension-polyfill/dist/browser-polyfill.js',
+        from: 'node_modules/webextension-polyfill-ts/dist/',
```
Why is this needed? 
I still see `lib/browser-polyfill.js` in the `extension/` dir post-build, although I would expect this change to break that (`node_modules/webextension-polyfill-ts/dist/` package doesn't seem to contain `browser-polyfill.js`).

aswinmprabhu replied:
This was just for testing.

poltak commented on src/activity-logger/background/log-page-visit.ts:
```diff
@@ -98,8 +99,41 @@ export default class PageVisitLogger {
+                ...analysisRes,
+            })
+            if (isPDF) {
+                getPDFFingerprint(tab.url).then(async fingerprint => {
```
I understand you haven't got the PDF fingerprint implementation yet, although the async flow of this is going to cause issues. `getPDFFingerprint` will run, then it will immediately move onto the following `this._createPage` call, as there's no `await` and that call is not inside the `.then()` block. And you'll end up calling `this._createPage` twice.

Instead you could do something like:

```ts
const fingerprint = isPDF ? await getPDFFingerprint(tab.url) : undefined
```

and remove the entire `.then()`, as the following `this._createPage` call will then take over. `undefined` over `null` here also as I think if it's `null` you'll end up telling IDB to store a `null` value for lots of pages. If it's `undefined`, IDB just skips that field.

poltak commented on src/activity-logger/background/pdffingerprint.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint eqeqeq: 0 */
+export default function getPDFFingerprint(pdfURL) {
+    const PDFJS = require('../../build/pdf')
+    PDFJS.GlobalWorkerOptions.workerSrc = '../../build/pdf.worker.js'
```
A few things about this module:
1. let's make all new modules `.ts` or `.tsx` modules (you can still write your standard JS stuff in those modules if you're new to TS)
2. I'm not sure if you're going to have an issue here with the relative paths. Those are relative to the source code structure although, once the ext is running in the browser post-build, that structure no longer exists and paths are now relative to the produced `extension/` dir. Have a look how we've done it using `extension.getURL` API for our other pdfjs uses: https://github.com/WorldBrain/Memex/blob/develop/src/page-analysis/content_script/extract-pdf-content.js#L9
3. There shouldn't be any reason to instantiate and return a `Promise` here. You can make the overall function `async`, then it implicitly returns a `Promise`. Though that assumes the rest of the pdfjs API is promise-based.

poltak commented on src/manifest.json:
```diff
@@ -4,6 +4,7 @@
     "version": "0.3.3",
     "offline_enabled": true,
     "description": "The bookmarking tool that works like your brain: Annotate & Full-Text Search anything you've seen on the web.",
+    "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'",
```
Probably need an explanation for this one too. Can you break this CSP down for me? We actually got kicked from Firefox addons a while back for releasing with a too open CSP, so we want to try and avoid that.

aswinmprabhu replied:
This content security policy is to get the `<script>` tags to work in the PDF.js viewer. The script tags are how I am adding the content scripts to the viewer. The 'unsafe-eval' directive allows inline execution of JS with script tags.

poltak replied:
`unsafe-eval` is the one that got us kicked off last time. I actually found the part in the MDN docs they state that:
http://memex.link/gCB4CjcUO/developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy

But from what I understand of the docs (and the Chrome-specific docs: https://developer.chrome.com/extensions/contentSecurityPolicy#relaxing-eval), `unsafe-eval` is **not** related to the inline execution of JS in script tags. 

According to the docs, it is possible, although it seems super painful.
Chrome: http://memex.link/yHOqrZVMK/developer.chrome.com/extensions/contentSecurityPolicy#JSExecution
MDN: http://memex.link/zJ1uYQjkq/developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy

Can you link me to where you are adding the content scripts in the PDF viewer, or doing inline JS, via `<script>` tags?

poltak commented on src/overview/results/components/PageResultItem.tsx:
```diff
@@ -14,6 +14,8 @@ const styles = require('./PageResultItem.css')
 
 export interface Props {
     url: string
+    PDFFingerprint: string
```
stick to `lowerCamelCase` for all standard variables

aswinmprabhu commented:
@poltak In order to reproduce the issues, open a PDF in the PDF.js viewer with the new button in the popup. Then try creating a comment from the sidebar. This throws the following errors in the console.
![err](https://user-images.githubusercontent.com/31558262/52568548-14e1d680-2e35-11e9-925a-bfe4cb82ed25.png)
These are coming from `sidebar.js`, which is the js file used in the sidebar iframe. The errors are not being thrown because of an RPC issue, but because `browser.tabs` is undefined [here](https://github.com/aswinmprabhu/Memex/blob/96123393e880bd9e9af9e3287f46375e8f492105/src/util/webextensionRPC.ts#L77). This is evident if you try to console log `browser.tabs` here.

poltak commented:
Thanks for addressing those review points so fast!

> These are coming from `sidebar.js`

By this do you mean the `sidebar.js` file that is output from the extension build? I cannot find any modules in the codebase with that name apart from some react view components. If it is, I recall @mohitsakhuja is going to remove that with his new work on the sidebar shadow DOM. Although I personally don't know what the status of that is yet... I'm a little worried about the overlap between these. As I talked about on slack, I'm feeling that we should wait until more is known about that work before attempting more on this feature, else we could just be setting ourselves up for unnecessary work in the near future.

Though there is still stuff to do with this work in the meantime. I mainly need answers to points 2 and 3 in my last comment. Like with the keyboard shortcuts work, these deps should not be here unless there's a good reason (not available on GH or NPM, for ex), and if there is a good reason we need to then think about how we'll include them in the codebase (ignore my last point 4 for now), make sure we include the licences, etc.

oliversauter commented:
> and if there is a good reason we need to then think about how we'll include them in the codebase

I have the slight worry that we are also doing double work in regards to generally implementing the PDF annotation viewer completely differently than Hypothes.is? Or is that worry unwarranted?

aswinmprabhu commented:
@poltak I got the PDF.js directories from their [official dist download page](https://mozilla.github.io/pdf.js/getting_started/#download). The github repo can be found [here](https://github.com/mozilla/pdf.js). I agree that deps should not be included like this. This was just for quick prototyping and testing.

oliversauter commented:
Let's pause this for now until we have the sidebar stuff all merged in. 
After that I think it would be good to make a status check to see what is necessary to implement this cleanly.

poltak commented:
@aswinmprabhu going by what we talked about on notion a while back, I thought you needed to use some fork of pdfjs which contained some extra functionality not present in standard pdfjs? If it is just standard pdfjs we already have `pdf-dist` as a Memex dep (what I believe is in your `src/build` dir), so you can remove that dir completely. I can't seem to find the viewer on NPM, though you can add the GH repo as a dep and remove that `src/web` dir.

For any future work, please just stick to NPM or GH deps (they're also going to be much easier for testing/prototyping than manually including source files; feel free to hit me up on slack if you're not sure). If you need to use some dep not available on those, then we need to discuss the best ways to include them in beforehand.

> I have the slight worry that we are also doing double work in regards to generally implementing the PDF annotation viewer completely differently than Hypothes.is? Or is that worry unwarranted?

@oliversauter Is the Hypothes.is PDF viewer available for use? Is it related to pdfjs in anyway? And will it save us time to just use that one instead of pdfjs viewer? These are definitely questions we need to answer before continuing on with implementation of this too

oliversauter commented:
> @oliversauter Is the Hypothes.is PDF viewer available for use? Is it related to pdfjs in anyway? And will it save us time to just use that one instead of pdfjs viewer? These are definitely questions we need to answer before continuing on with implementation of this too

Yeah its a PDF.js adaption too. Also they are in the mode of fixing some bugs on Firefox based on some changes to the PDF.js APIs. So we want to make use of those too. 
One of their devs wrote me: 
``On a related note, we're currently making some changes to the PDF
anchoring code in Hypothesis for compatibility with newer PDF.js
versions. See https://github.com/hypothesis/client/pull/889 and
earlier PRs for details.``

Not sure why we don't use it, @aswinmprabhu made some assessment on this. 

@aswinmprabhu You mentioned in this notion file that:
"Oh we are not using their version of pdf.js currently."

@aswinmprabhu aWhy don't we, what was the issue again for not using their version of PDF.js?

aswinmprabhu commented:
There is no particular reason as to why we don't use the hypothes.is version of PDF.js. I just went with the official version when I was prototyping. I have not yet found a case where I felt that the official version was causing the problem.

oliversauter commented:
But aren't we running now into the same issues Hypothes.is is doing righ now or are they completely different? 
Do you have an idea how we can solve #683 and this one, they seem to be related, no?

aswinmprabhu commented:
The shadowdom implementation of the sidebar fixed a lot of issues. So I went ahead and fixed a lot of these errors. Annotation based on pdf fingerprint works now. The comments and annotations are displayed on both the regular pdf and the pdfjs-viewer page. I also added a few things like the annotate pdf button on the sidebar.

oliversauter commented:
WHUUT :) will give it a test asap!

oliversauter commented:
ok gave it a go and it works now :) 
A couple of issues are still there:
1) when i go to the dashboard after having made some notes on the PDF, it does not load my results anymore. This is the error log: <img width="972" alt="screen shot 2019-02-24 at 16 54 20" src="https://user-images.githubusercontent.com/7870039/53302323-080dab80-3855-11e9-812f-f644e93df11d.png">
2) Also does not load the annotations on the URL Version after making annotations in the pdfjs-version:
<img width="851" alt="screen shot 2019-02-24 at 16 58 19" src="https://user-images.githubusercontent.com/7870039/53302371-6a66ac00-3855-11e9-9a76-3d18b3b671c1.png">

Steps to reproduce:
1) go to pdf, let it fetch
2) go to dashboard check if pdf is indexed
3) go back to pdf make comment on url version
4) open annotation viewer and make first annotation
5) go back to dashboard, reload it if necesessary > see "no results for this query" message



**Other remarks:**
- [ ] The annotation sidebar is behind the topbar of the PDF viewer (also saw that on other pages), can you make sure that the sidebar is always on top of everything else?
- [ ] where can I play around with the highlight color of the annotations? they seem very mild and almost not recognisable. I feel they should be separate from the website colors.
![screen shot 2019-02-24 at 16 53 02](https://user-images.githubusercontent.com/7870039/53302394-94b86980-3855-11e9-8f70-6634df214977.png)

aswinmprabhu commented:
These errors seem to only occur when the pdfjs-viewer is open when you are using the dashboard or the url version of the pdf. The rpc calls seem to be going to the pdfjs-viewer. I am not entirely sure why that is. I'll look into it. Meanwhile I fixed the sidebar issue. You can tweak the colors of the highlight [here](https://github.com/WorldBrain/Memex/blob/dd0acf424d732b0d21db7c74de5d3687cf1fa6f5/src/direct-linking/content_script/styles.css#L2).

oliversauter commented:
> You can tweak the colors of the highlight [here](https://github.com/WorldBrain/Memex/blob/dd0acf424d732b0d21db7c74de5d3687cf1fa6f5/src/direct-linking/content_script/styles.css#L2).

Can we have a separate class in there for the annotation colors. it seems as if changing them there would affect it in-page too much. (but first the other stuff we need to figure out, so that can go last.

aswinmprabhu commented:
I can't seem to figure out why these errors occur only when the pdfjs-viewer is open. The rpc calls seem to be going to the pdfjs-viewer. I need some help with this. @poltak How is the destination of the rpc call decided.

poltak commented:
> How is the destination of the rpc call decided?

If you're going by the `util/webextensionRPC` module, AFAIK it works like this:

From any script A, if you call `makeRemotelyCallable` on some function called `'f'`, any `remoteFunction('f')` call from any other script will set up another function to run that function `'f'` in script A.
There shouldn't be any other `makeRemotelyCallable` call on a function of the same name from any other script, or there's bound to be undesired behavior.

>  The rpc calls seem to be going to the pdfjs-viewer.

Where are you calling `makeRemotelyCallable` from?

Another thing that might help, just to better understand what's going on, would be to play around with the Web Ext APIs directly (`util/webextensionRPC` module just wraps around them). `runtime.sendMessage` is like our `remoteFunction` and `runtime.onMessage` event is like our `makeRemotelyCallable`. Docs:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage

oliversauter commented:
@aswinmprabhu Any progress on this one? Is there anything you struggle with that we can help you with?

poltak commented:
I have finished the main part of all the work I was working on recently, so I would like to help push this feature forward. Let us know what's the status @aswinmprabhu and what your plan is with it; if you don't have much time to work on it, we can continue on your work.

The main blockers I see at the moment, not directly related to the feature, but super important:
- the deps that have been pulled in via their source code: there needs to be a reason for doing this over using NPM or git repo based deps
- the `unsafe-eval` usage in the CSP: that will get us kicked off firefox addons, so we need to look at other ways

MartinLichtblau commented:
Hi @aswinmprabhu are you still there?

poltak commented:
Closing this in favour of #822. That branch is built on top of all of the work you did so far on this @aswinmprabhu :) Thanks a lot for the efforts with this feature!
